_posts/resources/2023-05-08-morello-linux-morelloie.md

```diff
@@ -0,0 +1,1098 @@
+
+// Set capability offset
+void *__builtin_cheri_offset_set(const void *cap, size_t offset);
+
+// Increment capability offset
+void *__builtin_cheri_offset_increment(const void *cap , size_t inc);
+
+// Set capability value (address)
+void *__builtin_cheri_address_set(const void *cap, size_t address);
+
+// Set upper 64 bits of a capability (will clear its tag)
+void *__builtin_cheri_copy_to_high(const void *cap, size_t hi);
+
+// Clear capability tag. Note: there isn't any tag_set function
+void *__builtin_cheri_tag_clear(const void *cap);
+
+// Apply bit mask to capability permissions (should be used to
+// clear certain permissions or make sure only certain permission are set)
+void *__builtin_cheri_perms_and(const void *cap, size_t perms);
+
+// Convert pointer to capability from another capability,
+// with null capability from zero semantics.
+void *__builtin_cheri_cap_from_pointer(const void *cap, size_t address);
+```
+{: .language-cpp}
+
+One notable use case is worth looking at: create a capability with given length with some required offset (i.e., given bounds) from a source capability that potentially has larger bounds. Consider the following example:
+
+```
+#include <stdio.h>
+
+int main() {
+    const void *csp = __builtin_cheri_stack_get();
+    void *cap = __builtin_cheri_offset_increment(csp, -64);
+    cap = __builtin_cheri_bounds_set(cap, 64);
+    cap = __builtin_cheri_offset_set(cap, 16);
+    printf("cap = %#p\n", cap);
+    return 0;
+}
+```
+{: .language-cpp}
+
+Here we derive a capability from `CSP` and set its base to `csp.address - 64` and its length is 64 bytes while its offset is 16 bytes (with respect to the base value).
+
+It is also important to understand difference between the `__builtin_cheri_address_set` and `__builtin_cheri_cap_from_pointer` builtins. The former will set value of the capability potentially creating an invalid capability while the latter will also create a `NULL` capability if the address is zero.
+
+#### Sealing and unsealing capabilities
+
+```
+// Seal function pointer (creating a sentry)
+// On Morello, this creates RB-sealed capability.
+void *__builtin_cheri_seal_entry(const void *cap);
+
+// Seal a capability using another sealer capability
+void *__builtin_cheri_seal(const void *cap, const void *sealer);
+
+// Conditionally seal a capability using another sealer capability
+void *__builtin_cheri_conditional_seal(const void *cap, const void *sealer);
+
+// Unseal capability using another unsealing capability
+void *__builtin_cheri_unseal(const void *cap, const void *sealer);
+
+// Set capability value to the object type of another capability
+void *__builtin_cheri_cap_type_copy(const void *cap, const void *otype);
+```
+{: .language-cpp}
+
+Note: there is no builtin for LPB and LB object types, you will need to use inline assembly for this:
+
+```
+inline __attribute__ ((naked))
+void *__morello_seal_lpb(void *cap)
+{
+__asm__ volatile(
+"   seal    c0, c0, lpb\n"
+"   ret\n"
+);
+}
```

Comment:
```suggestion
inline __attribute__ ((naked))
void *__morello_seal_lpb(void *cap)
{
    void *ret;

    __asm__ ("seal %0, %1, lpb" : "=C"(ret) : "C"(cap));
    return (ret);
}
```
is the much less gross way of doing it (not just conceptually; it gives the compiler more flexibility too).